### DISCUSSION TITLE
Webhook with custom payload

### DISCUSSION THREAD

**Discussion by orjanskaar** [vendurehq/vendure#1739] [Q&A]
Hello

I have a working webhook plugin from [Pinelab Studio](https://github.com/Pinelab-studio/pinelab-vendure-plugins) fireing on "StockMovementEvent".
This plugin sends an empty body as stated in their [readme](https://github.com/Pinelab-studio/pinelab-vendure-plugins/blob/master/packages/vendure-plugin-webhook/README.md).

Is there a way to make a plugin to send with som "useful" information like:
-which variant id has triggered the change in stock
-Balance before change
-Balance after change
etc..

I have looked into the "[RequestContext](https://www.vendure.io/docs/typescript-api/request/request-context/#requestcontext)" section of the documentation, but im unsure if that is the correct path and also
unsure if this can provide the information im looking for.

The usecase for this (in my case) is that i have an exterrnal product database (zettle) for sales in a physical store which shares a common product inventory with vendure
With sales in the zettle app i have managed to update stock in vendure as there is the required information in the body of the request (variant id etc), but for sales in my webshop i currently have a no way to update stock in the zettle database

Thank you for any input :)

**Comment by michaelbromley** [comment]
Hi,

I've not personally used that plugin, but maybe if I summon @martijnvdbrug he might be able to advise?

**Reply by orjanskaar** [reply]
Hi

In retrospect i see that i should have included that in my first post...
I reached out to [martijnvdbrug](https://github.com/martijnvdbrug) in May and he [suggested ](https://github.com/Pinelab-studio/pinelab-vendure-plugins/issues/30) that I open a PR as it shouldn't be too complicated to implement in the existing code.

However i'm shure i've would have asked the same question in here in the end...
But perhaps he could feel encourraged if he is summoned by someone like you?

Thank you for your reply :)

**Comment by martijnvdbrug** [comment]
@orjanskaar could you create an issue for this here https://github.com/Pinelab-studio/pinelab-vendure-plugins

It shouldnt be too hard to implement some kind of strategy pattern so everyone can define his own message body.
 It could be a few weeks before I get to it, so any PR's are also welcome

**Reply by orjanskaar** [reply]
Hi 
I opened one some time ago. I will continue the discussion [there](https://github.com/Pinelab-studio/pinelab-vendure-plugins/issues/30)

**Reply by martijnvdbrug** [reply]
@orjanskaar Yep, I missed that one! But, [`vendure-plugin-webhook` v3.12.0](https://www.npmjs.com/package/vendure-plugin-webhook) now allows you to define a POST body (and headers) via a custom function:
```ts
WebhookPlugin.init({
    httpMethod: 'POST',
    events: [ProductEvent, ProductVariantChannelEvent, ProductVariantEvent],
    requestFn: (event) => {
      return {
        headers: { test: '1234' },
        body: JSON.stringify({ createdAt: event.createdAt }),
      };
    },
  }),
```

**Reply by orjanskaar** [reply]
Hello 

Thank you for you reply!

With the following i got the data I needed

```typescript
WebhookPlugin.init({
    httpMethod: 'POST',
    events: [ ProductVariantEvent ],
    requestFn: (event) => {
      return {
        body: JSON.stringify({ payload: event.entity }),
      };
    },
  }),
